internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -787,10 +797,19 @@ func (d *DeletingComposedResourceGarbageCollector) GarbageCollectComposedResourc
 
 // UpdateResourceRefs updates the supplied state to ensure the XR references all
 // composed resources that exist or are pending creation.
-func UpdateResourceRefs(xr resource.ComposedResourcesReferencer, desired ComposedResourceStates) {
+func UpdateResourceRefs(xr xresource.Composite, desired ComposedResourceStates) {
+	namespaced := xr.GetNamespace() != ""
 	refs := make([]corev1.ObjectReference, 0, len(desired))
 	for _, dr := range desired {
 		ref := meta.ReferenceTo(dr.Resource, dr.Resource.GetObjectKind().GroupVersionKind())
+
+		// If the XR is namespaced it can only compose resources in its own
+		// namespace. Its OpenAPI schema won't allow including a namespace in
+		// its resourceRefs.
+		if namespaced {
+			ref.Namespace = ""
```

Comment:
what would it mean if a namespace was specified on the desired resource? are we essentially dropping that information here?

Reply:
We handle resources composed by namespaced XRs specially in a few places:

* `ObserveComposedResources` always looks for composed resources in the XR's namespace
* `RenderComposedResourceMetadata` sets the namespace of composed resource to the XR's namespace.
* `UpdateResourceRefs` omits the namespace when recording resource refs to composed resources

So yeah we drop the information, but it's not possible for the resource to be created in another namespace. If a function returns a specific namespace for a desired resource, Crossplane will overwrite that with the XR's namespace.